packages/amplify_core/lib/src/amplify_class.dart

```diff
@@ -90,21 +90,43 @@ abstract class AmplifyClass {
       );
     }
 
-    final AmplifyConfig amplifyConfig;
+    late AmplifyConfig amplifyConfig;
     try {
-      final json = jsonDecode(configuration) as Map;
-      amplifyConfig = AmplifyConfig.fromJson(json.cast());
-    } on FormatException catch (e) {
-      throw AmplifyException(
-        'The provided configuration is not a valid json. Check underlyingException.',
-        recoverySuggestion:
-            'Inspect your amplifyconfiguration.dart and ensure that the string is proper json',
-        underlyingException: e,
-      );
+      try {
+        final json = jsonDecode(configuration) as Map;
+        amplifyConfig = AmplifyConfig.fromJson(json.cast());
+      } on Object catch (e) {
+        throw ConfigurationException(
+          'The provided configuration is not a valid json. Check underlyingException.',
+          recoverySuggestion:
+              'Inspect your amplifyconfiguration.dart and ensure that the string is proper json',
+          underlyingException: e,
+        );
+      }
+      await configurePlatform(configuration);
+      _configCompleter.complete(amplifyConfig);
+    } on ConfigurationException catch (e, st) {
+      // Complete with the configuration error and reset the completer so
+      // that 1) `configure` can be called again and 2) listeners registered
+      // on `asyncConfig` are notified of the error so they can handle it as
+      // appropriate. For example, the Authenticator listens for `asyncConfig`
+      // to complete before updating the UI.
```

Comment:
Appreciate the comments 👍 

My understanding is that we are sort of treating `ConfigurationException` as a fatal exception. That  is, it is thrown in scenarios where we know plugins will not be functional (missing or invalid config). Configure will have to be called again to get plugins into a functional state.

If there are scenarios where an exception may leave plugins in a functional state (not sure what scenario this would be. I guess the hosted UI issues you are resolving?), `ConfigurationException` would not be appropriate. Another exception should be thrown, which will allow `asyncConfig` to complete. Is that understanding correct?

It might be worth adding some of this info to the doc comment for `ConfigurationException`, or maybe the getter for  `asyncConfig`. I could imagine `ConfigurationException` being used incorrectly if it isn't very clear how it should be used.

Reply:
I almost made it `ConfigurationError extends Error` since it really is unrecoverable. Do you think that would be more clear?

Reply:
Updated to `ConfigurationError` and expanded on its meaning in its docs.